www/blog/2025-06-24-efficient-delta-updates.mdx

```diff
@@ -0,0 +1,166 @@
+
+[^moseley-delta]: [Torizon Blog: Over-The-Air Updates | Choose the best Linux OS Image Update Model](https://www.torizon.io/de/blog/ota-best-linux-os-image-update-model)
+
+Looking at the data, the difference between block-based and file-based diffing is not very large. However, if we look at delta compression (*Xdelta*) in comparison, we find that delta compression can be more than twice as efficient as block-based and file-based diffing. Recall that individual package updates, are essentially file-based updates. Hence, delta compression can be even more efficient than updating individual packages. This efficiency gain is also not just a result of the fact that it is a static delta update approach. Given the negligible sizes of block and file indices, static variants of block-based and file-based diffing would not be much more efficient than their dynamic counterparts. While delta compression is the most efficient approach, it requires pre-computing updates and it is also the most complex to implement and operate. If you need to reduce the size of updates to an absolute minimum, delta compression is probably the right choice for rolling updates. The results presented here are typical for Rugix's and Mender's static delta updates, as they both use Xdelta internally.
+
+### Major Upgrades
+
+Let's now turn to major upgrades. Major upgrades are not as frequent and typically contain a lot of significant changes. As for rolling updates, we also conducted benchmarks for the different update cadences, but instead of updating within a Debian version in each step, we update from Debian Bullseye to Bookworm in each step and then average the results. This represents a major update being done with the specific cadence but at different points in time. Here are the results:
+
+<PlotMajorUpdates />
```

Comment:
I'm still wrapping my head around the graph, but the visualization looks off because the "no delta" updates "looks better" (as the bar is smaller).

I'm sure the figures are technically correct, but just look misleading as a user. I would expect the delta algorithms to still be "less than the no delta update", so it might just warrant rethinking how it is visualized.

Reply:
Actually, the interpretation that delta updates are worse is correct. So, no delta updates are indeed better. This is because you get overhead with delta updates and that overhead does not pay off when there are too many changes. Was that just the first impression and did it become clear after reading on? I tried to explain it. It is indeed a bit unexpected. :D

Reply:
Ok, I didn't read get around to reading the paragraph before posting the comment.

Though it would seem that it would be fairly trivial for an image maintainer to identify whether a delta-update makes sense for a specific release (e.g. when the changes are too significant to warrant using it), and revert to just sending the new image in full.